lsp-mode.el

```diff
@@ -3146,6 +3146,10 @@ The method uses `replace-buffer-contents'."
                            'lsp--apply-text-edit)))
         (unwind-protect
             (->> edits
+                 (mapc (lambda (edit)
```

Comment:
having this unconditionally seems to be opinionated and this might be a problem in mixed Windows/Linux teams with coding conventions. Are there functional impact when the ^Ms are present?  As a side note it is possible to hide ^M https://stackoverflow.com/questions/730751/hiding-m-in-emacs .

Reply:
Hiding `\r` in Emacs doesn't seem good to me. If we should use that code, it should be in user config file.

With powershell language server, whenever user send `\n` and ask for formatting, the server will return `\r\n` as it's the standard EOL in Windows.
Now in Emacs, when the file coding is set to `-dos`, the `\r\n` will automatically converted to `\n` (at least that happens when you call `(buffer-string)`). 

So we have an Emacs specific issue here:
- Whenever we request formatting, powershell server will answer with `\r\n`.
- Now go to Emacs, even if the file encoding is set to `-dos`, if we're replacing text EOL with `\r\n`, the `\r` will be treated as extra character, and `^M` will be shown.
- Worse, next time we request formatting with `\r\n`, the server will return `\r\r\n` and inflate the buffer indefinitely.

Once the buffer is in its correct encoding (`-dos`, `-linux` and `-mac`), the newline will always be represented as `\n` instead of `\r\n`, so removing `\r` would be fine IMO.
However, if this change is risky, can you suggest a better scoped change for this?

Reply:
> Hiding \r in Emacs doesn't seem good to me.

Yes, I am not suggesting it. Just mentioning that if there is no functional impact but only visual there is an option to hide it. 

> if we're replacing text EOL with \r\n, the \r will be treated as extra character, and ^M will be shown.

Why do we need to do that?

> Worse, next time we request formatting with \r\n, the server will return \r\r\n and inflate the buffer indefinitely.

Can we avoid that? This sounds like an issue in lsp-mode.

> Once the buffer is in its correct encoding (-dos, -linux and -mac), the newline will always be represented as \n instead of \r\n, so removing \r would be fine IMO.

Can you elaborate on that?

> However, if this change is risky, can you suggest a better scoped change for this?

We may have a flag to enable it and enable it for powershell buffers via buffer local.

Reply:
> Why do we need to do that?

That's what powershell language server does. `lsp-mode` sends EOL as `\n` and the language server will answer that with `\r\n` as it's standard EOL in Windows. Now when we try to replace the current buffer with what language server sent, `\r` will be left there alone, visible.

> Can we avoid that? This sounds like an issue in lsp-mode.

That why this change try to remove `\r` from text returned from language server, to match with what Emacs expects (I guess).

>Can you elaborate on that?

For example, a buffer contains "a\r\n". If we encode it using `utf-8-dos`, we receive "a\n" with `buffer-string`. If we encode it using `utf-8-unix`, we get "a\r\n".
Both cases, the EOL is just "\n". After we fix the buffer encoding and do text replacement, any extra "\r" will be visible and should be safely removed.

Reply:
So, what you are saying is that removing \r is practically noop since it will be either removed by emacs if it is not needed or automatically added if needed?

Reply:
> So, what you are saying is that removing \r is practically noop since it will be either removed by emacs if it is not needed or automatically added if needed?

That's not I meant. Sorry.
What I mean is that in Emacs, once we encoding the buffer correctly, the EOL will always be `\n`, having extra `\r`s around are mostly redundant and distracting.

I'm curious, did we convert `\n` to `\r\n` when sync document with server in Windows?

Reply:
> That's not I meant. Sorry.
What I mean is that in Emacs, once we encoding the buffer correctly, the EOL will always be \n, having extra \rs around are mostly redundant and distracting.

What will happen if the user does want dos encoding? Will he get \r\n?

> I'm curious, did we convert \n to \r\n when sync document with server in Windows?

AFAIK no. 

@emacs-lsp/lsp-mode does anybody have more knowledge regarding the matter?

Reply:
> What will happen if the user does want dos encoding? Will he get \r\n?

The replaced text and file encoding is separated matter. 
For example, if the file encoding is `-dos`, and user save string "a\n" into file. It will be saved as "a\r\n" and user will only see "a\n" with `buffer-string`.
Similarly, if the file encoding is `-unix`, and user save string "a\n" into file, It will be saved as "a\n" and user will only see "a\n" as `buffer-string`

Now since language server will try to replace "\n" with "\r\n", the `buffer-string` will become something like "a\r\n" in both `-dos` and `-unix` encoding. And respectively, "a\r\r\n" and "a\r\n" in file content.
You can see that the "\r" is redundant as it interfere with the current file encoding

Reply:
I tested mssql server with that change and it solved some issues. I will wait for some time before merging it so others can share their opinion. Meanwhile, can you add a variable to disable this behaviour (don't use defcustom but defvar).

Reply:
Added